### DISCUSSION TITLE
Datacap timing

### DISCUSSION THREAD

**Discussion by kernelogic** [filecoin-project/lotus#5660] [Storage Provider] [answered]
Have a couple questions when making offline verified deals with folks

1. When does the datacap get deducted?
2. When does the datacap get returned for failed deals?
3. It looks like when the miner is batch publish deals, even if there's datacap left for some of the deals, but if there's one deal that reach the limit, the whole batch fails with error "deal failed: (State=11) error calling node: publishing deal: GasEstimateMessageGas error: estimating gas used: message execution failed: exit 17, reason: failed to add verified deal for client:"

**Comment by kernelogic** [comment]
I think ideally the datacap should be "locked" upon proposing the deal, not when the deal is published.

**Comment by jennijuju** [comment]
@kernelogic its detuced here https://github.com/filecoin-project/specs-actors/blob/845089a6d2580e46055c24415a6c32ee688e5186/actors/builtin/market/market_actor.go#L254

**Comment by dd45e640b42e6da7da96faee3996ef7c** [comment]
@kernelogic the deal is totally offchain until it is published. locking up the data cap before hand would be a soft lock on the client side only and there would be no way for a miner to test if the cap is locked or not.

if we lock it a miner could, kind of, hold the datacap hostage by not publishing it if we do not implement a "reverse the lock" command (it isn't locked anyways since the chain doesn't know about it)

a miner needs to take care of not publishing already taken data caps! the current miner deal handling logic seems not to take care of that - maybe we should start an issue.

[edit]

there is already an issue in the pipeline for that, i closed mine

https://github.com/filecoin-project/lotus/issues/5661

**Comment by jennijuju** [comment, marked as answer]
fixed https://github.com/filecoin-project/go-fil-markets/blob/32f7f2ed36dcb71473812905d0[…]7ee077d824/storagemarket/impl/providerstates/provider_states.go
